libs/math/include/math/activation_functions/softmax.hpp

```diff
@@ -99,9 +99,20 @@ void Softmax(ArrayType const &array, ArrayType &ret, typename ArrayType::SizeTyp
   {
     details::Softmax2DImplementation(array, ret, axis);
   }
+  else if ((array.shape().size() == 3) && (ret.shape().size() == 3) && (axis == 0 || axis == 1))
+  {
+    ArrayType tmp_ret = ret.Slice(0, 2).Copy().Squeeze();
+    for (size_t i = 0; i < array.shape()[2]; i++)
+    {
+      details::Softmax2DImplementation(array.Slice(i, 2).Copy().Squeeze(), tmp_ret, axis);
+      ret.Slice(i, 2).Assign(tmp_ret);
+    }
+  }
   else
```

Comment:
add another else if to catch the case for 3D with axis == 2;
in this case you should use Slices

Reply:
I am not sure for now if we should implement the activation on batch dimension. It is better that we leave it till someone implemented multi dimension softmax.